Mathlib/Tactic/Linter/Lint.lean

```diff
@@ -211,6 +211,51 @@ initialize addLinter cdotLinter
 
 end CDotLinter
 
+/-!
+# The `lambdaSyntax` linter
+
+The `lambdaSyntax` linter is a syntax linter that flags uses of the symbol `λ` to define anonymous
+functions, as opposed to the `fun` keyword. These are syntactically equivalent; mathlib style
+prefers the latter as it is considered more readable.
+-/
+
+/--
+The `lambdaSyntax` linter flags uses of the symbol `λ` to define anonymous functions.
+This is syntactically equivalent to the `fun` keyword; mathlib style prefers using the latter.-/
+register_option linter.lambdaSyntax : Bool := {
+  defValue := true
+  descr := "enable the `lambdaSyntax` linter"
+}
+
+namespace LambdaSyntaxLinter
+
+/--
+`findLambdaSyntax stx` extracts from `stx` all syntax nodes of `kind` `Term.fun`. -/
+partial
+def findLambdaSyntax : Syntax → Array (Syntax)
+  | stx@(.node _ kind args) =>
+    let dargs := (args.map findLambdaSyntax).flatten
+    match kind with
+      | ``Parser.Term.fun => dargs.push stx
+      | _ =>  dargs
+  |_ => #[]
+
+@[inherit_doc linter.lambdaSyntax]
+def lambdaSyntaxLinter : Linter where run := withSetOptionIn fun stx ↦ do
+    unless Linter.getLinterValue linter.lambdaSyntax (← getOptions) do
+      return
+    if (← MonadState.get).messages.hasErrors then
+      return
+    for s in findLambdaSyntax stx do
+      if let .atom _ "λ" := s.getArgs[0]! then
+        Linter.logLint linter.lambdaSyntax s m!"\
+        Please use 'fun' and not 'λ' to define anonymous functions.\n\
+        The latter syntax has been deprecated in mathlib 4."
```

Comment:
Note that this approach does not allow:
```
example : ℕ → ℕ := set_option linter.lambdaSyntax false in λ _ ↦ 0
```
which would require walking the infotree instead.

Reply:
Good point! Not supporting this seems fine to me, but I have added a test to this effect.